service/history/shard/context.go

```diff
@@ -216,10 +214,42 @@ func (s *contextImpl) GenerateTaskIDs(number int) ([]int64, error) {
 	return result, nil
 }
 
-func (s *contextImpl) GetTransferMaxReadLevel() int64 {
+func (s *contextImpl) GetQueueMaxReadLevel(category persistence.HistoryTaskCategory, cluster string) persistence.HistoryTaskKey {
+	switch category.Type() {
+	case persistence.HistoryTaskCategoryTypeImmediate:
+		return s.getImmediateTaskMaxReadLevel()
+	case persistence.HistoryTaskCategoryTypeScheduled:
+		return s.updateScheduledTaskMaxReadLevel(cluster)
+	default:
+		s.logger.Fatal("unknown history task category", tag.Dynamic("category-type", category.Type()))
```

Comment:
Ideally we should avoid Fatal loggic this deep in the call hierarchy. It could return (key, error). However this might not be feasible to propagate. Consider this option and propagate error if applicable